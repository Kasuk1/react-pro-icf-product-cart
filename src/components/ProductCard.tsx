import React, { createContext, CSSProperties } from 'react';
import { useProduct } from '../hooks/useProduct';
import {
  IProductContext,
  IProduct,
  IOnChangeArgs,
  IInitialvalues,
} from '../interfaces/interfaces';
import styles from '../styles/styles.module.css';
import { IProductCardHandlers } from '../interfaces/interfaces';

export const ProductContext = createContext({} as IProductContext);
const { Provider } = ProductContext;
export interface IProductCard {
  //children?: ReactElement | ReactElement[];
  children: (args: IProductCardHandlers) => JSX.Element;
  product: IProduct;
  className?: string;
  style?: CSSProperties;
  onChange?: (args: IOnChangeArgs) => void;
  value?: number;
  initialValues?: IInitialvalues;
}

export const ProductCard = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValues,
}: IProductCard) => {
  const {
    counter,
    increaseBy,
    maxCount,
    isMaxCountReached,
    reset,
  } = useProduct({
    product,
    onChange,
    value,
    initialValues,
  });

  return (
    <Provider
      value={{
        counter,
        increaseBy,
        product,
        maxCount,
      }}
    >
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          count: counter,
          maxCount: initialValues?.maxCount,
          isMaxCountReached,
          reset,
          increaseBy,
          product,
        })}
      </div>
    </Provider>
  );
};
