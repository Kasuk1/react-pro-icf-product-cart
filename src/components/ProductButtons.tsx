import React, { CSSProperties, useCallback, useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';

export interface IProductButtons {
  className?: string;
  style?: CSSProperties;
}

export const ProductButtons = ({ className, style }: IProductButtons) => {
  const { increaseBy, counter, maxCount } = useContext(ProductContext);

  const isMaxReached = useCallback(() => {
    return !!maxCount && counter === maxCount;
  }, [counter, maxCount]);

  return (
    <div className={`${styles.buttonsContainer} ${className}`} style={style}>
      <button className={styles.buttonMinus} onClick={() => increaseBy(-1)}>
        -
      </button>
      <div className={styles.countLabel}>{counter}</div>
      <button
        className={`${styles.buttonAdd} ${
          isMaxReached() ? styles.disabled : ''
        }`}
        onClick={() => increaseBy(1)}
        disabled={isMaxReached()}
      >
        +
      </button>
    </div>
  );
};
