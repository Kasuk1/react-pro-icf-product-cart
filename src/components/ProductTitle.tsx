import React, { CSSProperties, useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';

export interface IProductTitle {
  title?: string;
  className?: string;
  style?: CSSProperties;
}

export const ProductTitle = ({ title, className, style }: IProductTitle) => {
  const { product } = useContext(ProductContext);

  return (
    <span className={`${styles.productDescription} ${className}`} style={style}>
      {title ? title : product.title}
    </span>
  );
};
