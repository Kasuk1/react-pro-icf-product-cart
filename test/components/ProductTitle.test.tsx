import * as React from 'react';
import renderer from 'react-test-renderer';
import { ProductCard, ProductTitle } from '../../src/components';

import { product1 } from '../data/products';

describe('<ProductTitle /> Tests', () => {
  test('should show the component correctly with the title personalized', () => {
    const wrapper = renderer.create(<ProductTitle title="Testing title" />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('should show the component with product title', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>{({}) => <ProductTitle />}</ProductCard>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
