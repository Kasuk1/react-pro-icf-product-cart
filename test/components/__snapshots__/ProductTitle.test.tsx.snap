// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductTitle /> Tests should show the component correctly with the title personalized 1`] = `
<span
  className="productDescription undefined"
>
  Testing title
</span>
`;

exports[`<ProductTitle /> Tests should show the component with product title 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined"
  >
    Coffe Mug - No Image
  </span>
</div>
`;
